### DISCUSSION TITLE
thin lto brakes corefreq

### DISCUSSION THREAD

**Discussion by LethalManBoob** [cyring/CoreFreq#459] [Q&A] [answered]
cachyos-lto kernel causes this to happen someone reported they were unable to boot.
![image](https://github.com/cyring/CoreFreq/assets/30273211/7b120bb4-6974-4152-8ba7-fc6d0609ec49)

**Comment by cyring** [comment]
Indeed I can see bad connection issues with the backend:
* SysGate is Off
* No Vcore, neither temperature reported
* Caches and probably others are not displayed


CachyOS was a best effort to support but this kernel mod is too patched for a long term programming environment. 
Thus your guys have to troubleshoot more than that and will be kind to verify or provide the kernel log to investigate CachyOS Patches which may conflict with the Linux mainstream kernel.

**Comment by LethalManBoob** [comment]
Turns out the LTO THIN and rt brakes dkms
![image](https://github.com/cyring/CoreFreq/assets/30273211/5a85615d-17dd-45a0-9deb-ba9e5215a19e)

**Comment by cyring** [comment]
> Turns out the LTO THIN and rt brakes dkms ![image](https://user-images.githubusercontent.com/30273211/267352321-5a85615d-17dd-45a0-9deb-ba9e5215a19e.png)

Regarding the AUR you may get a little help from @inglor but not sure if CachyOS is supported by package.
If not then I would suggest you manually clone or download _CoreFreq_ source code, next build and run straight from the project directory. See _CoreFreq_ [`Readme.md`](https://github.com/cyring/CoreFreq#build)
However make sure before the kernel development prerequisites of _CoreFreq_ are well installed with your CachyOS: gcc compiler, make, kernel headers, glibc. These are usually pulled by the dkms package, even if you don't use it -;)

**Comment by inglor** [comment]
@LethalManBoob you might have a better support if you were using any of the Arch Linux kernels from the repos. Can you try with the default Arch Linux real-time kernel to check if this is irelevant of CachyOS patches?

Otherwise you probably need to ask in CachyOS support or something.

**Comment by LethalManBoob** [comment]
> @LethalManBoob you might have a better support if you were using any of the Arch Linux kernels from the repos. Can you try with the default Arch Linux real-time kernel to check if this is irelevant of CachyOS patches?
> 
> Otherwise you probably need to ask in CachyOS support or something.

It can build on the base aur linux-rt but not cachyos-rt

**Comment by inglor** [comment]
@LethalManBoob check with the maintainers of the cachyos-rt - probably one of their patches is breaking this.

**Comment by cyring** [comment]
@LethalManBoob 
Have a look to these previous solved issues:
* [Failure to build dkms modules](https://github.com/cyring/CoreFreq/issues/435#issuecomment-1536942645)
* [DKMS compilations fails on 6.3](https://github.com/cyring/CoreFreq/issues/422#issuecomment-1472724519)

**Comment by ptr1337** [comment]
> @LethalManBoob check with the maintainers of the cachyos-rt - probably one of their patches is breaking this.

He should provide the log of the failed compilation.
Also we are running a 6.5.3 RT on not 6.3. Not sure if this makes a difference, but we need a log.

Edit:
Also the RT Kernel does not differ much from the default "linux-cachyos". Only the recent RT patchset is applied.

**Comment by csutcliff** [comment]
My experience is that it works fine with normal kernels (including linux-cachyos) and builds fine via DKMS for the LTO variants but as soon as I boot an LTO kernel with the corefreq module enabled I cannot even log in. SDDM launches but after entering my password I get about 20 seconds of black screen before being dumped back at SDDM. If I switch to another TTY and try to login to the console it hangs after I enter username and password, never proceeds to a shell prompt. As I can't login via gui or console when this happens I don't know how to debug or get logs.

If I disable the corefreq module the LTO kernels otherwise run fine.

Has anyone got any ideas how to debug further?

**Comment by gel-crabs** [comment]
I've been using thin LTO with CoreFreq for a long time with no issues, I'm currently on cachyos 6.6-rc1 built with Clang and thin LTO. I'm almost certain this is a bug which was reported previously. It is fixed by rebuilding your kernel with CONFIG_FORTIFY_SOURCE disabled.

See https://github.com/cyring/CoreFreq/issues/422#issuecomment-1483790909

**Comment by ptr1337** [comment]
> I've been using thin LTO with CoreFreq for a long time with no issues, I'm currently on cachyos 6.6-rc1 built with Clang and thin LTO. I'm almost certain this is a bug which was reported previously. It is fixed by rebuilding your kernel with CONFIG_FORTIFY_SOURCE disabled.
> 
> See [#422 (comment)](https://github.com/cyring/CoreFreq/issues/422#issuecomment-1483790909)

This still seems not for me for a solution, to disable FORTIFY_SOURCE globally for our config, just because corefreq has issues with.
There should be further debugged, why FORTIFY_SOURCE in combination with clang is breaking it or FORTIFY_SOURCE should be completely disabled in the DKMS module, if clang is used.

Maybe reporting a bug to llvm/clang would be also worth to get some further information.
https://github.com/clangbuiltlinux/linux --> there should be quite easy to get some professional help for this issue, but not sure if they support out of tree modules.

Edit:
But using the GCC kernel should be also enough. The LTO Kernel is just some little extra, if people want to use it, but not our default.

**Comment by cyring** [comment]
Concerning `6.6.0-rc1`, no issue to build and run _CoreFreq_

**Comment by LethalManBoob** [comment]
cachyos-rt-lto would not boot with corefreq-client,dkms,server installed just a heads up.

**Reply by cyring** [reply]
Thanks for this. 

Can you post the outputs of a manual _CoreFreq_ build. I mean `make clean all`  

Once built, run _CoreFreq_ and please provide the kernel log of the driver. But also the Deamon output started in Debug mode like `corefreqd -d` and Cli outputs `corefreq-cli -s -n -m -n -M -n -B -n -k`

Because the _CoreFreq_ prerequisite `CONFIG_CACHY` is unexpectedly missing from `config` file in [linux-cachyos-rt](https://aur.archlinux.org/packages/linux-cachyos-rt), I would like also to verify your `/proc/config.gz`

**Reply by ptr1337** [reply]
`CONFIG_CACHY` gets enabled via the PKGBUILD

**Reply by cyring** [reply]
But not found within this [PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=linux-cachyos-rt)

**Reply by ptr1337** [reply]
https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=linux-cachyos-rt#n307

**Reply by ptr1337** [reply, marked as answer]
I did not test corefreq yet on RT, but I could imagine that the RT or THIN LTO is breaking corefreq with it.
We also don't suggest to use custom modules with Thin LTO or other custom stuff, which is not tested OOB.

**Reply by cyring** [reply]
@ptr1337 : Thanks for your explanations.

@LethalManBoob : this will be marked as the answer.
